### DISCUSSION TITLE
Troubleshooting new build....

### DISCUSSION THREAD

**Discussion by TactileDesign** [kriswilk/buzzah#21] [Q&A] [answered]
I have everything connected (albeit with just one motor) - processor is programmed as directed with files from download link, result:
Processor flashes red light apprx every 5 seconds, all other boards show green lights, but no buzzing of the LRA...  I am wondering if they might have sent me the wrong motors for Digikey?

First tried 1 driver board + 1 motor, then hooked all 8 boards with only 1 motor connected , still no buzzing.  Have tried 2 different motors - same result.   I tried looking at the output of the driver boards with my scope and got a 5 volt, 60 Hz wave (maybe a ground loop problem there between my USB supply and scope?).  Do all 8 motors need to be connected in order for system to function?  Any ideas?

**Comment by kriswilk** [comment]
The code will use whichever drivers are connected and will ignore empty ports. So you should be able to test with even a single driver/actuator.

The red light on the QT Py board suggests that the code isn't running properly. Someone else was having trouble yesterday... I'd start by referring to my [advice in that thread](https://github.com/kriswilk/buzzah/discussions/5#discussioncomment-9799835).

If you (re)install CircuitPython, the required "lib" folder' and `code.py` and still have trouble, the next step would be to connect the hardware to a PC and watch the "serial console", which is a text output from the QT Py board. It will likely show one or more errors and could help to diagnose the cause of the problem.

To see the serial console output, refer to the following page...it suggests a program called "Mu", which is a simple code editor and serial console all in one package.

https://learn.adafruit.com/welcome-to-circuitpython/kattni-connecting-to-the-serial-console

Let me know your progress...

**Reply by TactileDesign** [reply]
Thanks Kris, here is what I have done

1.  reloaded operating sys on rp as you suggested - no change
2.  downloaded and loaded older python operating sys - no change (my idea was that your code page said it was from 10 months ago, and maybe something changed in newer version of python)
3. I noticed that the lib and code files were surviving on the RP even after reset - so I reformatted the RP chip to clean out old files and repeated steps 1 and 2 - still no effect  (I hope I didnt loose anything in the formatting process!)
4. Installed Mu.  Plugged RP into computer with MUX and driver board attached.   RP currently running Python vs from Oct 23. 
5. Monitored the serial output and got this error message:
code.py output:
Traceback (most recent call last):
  File "code.py", line 78, in <module>
  File "adafruit_drv2605.py", line 93, in __init__
  File "adafruit_tca9548a.py", line 62, in try_lock
OSError: [Errno 19] No such device

Does this give you  any insights?
Thanks for your help!

**Reply by TactileDesign** [reply]
In case this helps...  ran Mu serial again and got slightly different error:
code.py output:
Traceback (most recent call last):
  File "code.py", line 78, in <module>
  File "adafruit_drv2605.py", line 93, in __init__
  File "adafruit_tca9548a.py", line 62, in try_lock
OSError: [Errno 116] ETIMEDOUT

**Reply by kriswilk** [reply]
Perhaps you can post a photo (or photos) showing your actual boards and connections? Looks like the code is having trouble connecting to the mux and/or driver board.

**Reply by TactileDesign** [reply]
will send pic in a moment...   
changed out cables and reloaded most recent python code and now get this error:

code.py output:
Traceback (most recent call last):
  File "code.py", line 1, in <module>
ValueError: incompatible .mpy file

Code done running.

**Reply by TactileDesign** [reply]
Here is a pic 

I have one more RP I ordered from Digitech as a backup.  Is it time to give up on this one and start the process anew with my last hope RP??  Thanks again for your patience and help!
![IMG_7851](https://github.com/kriswilk/buzzah/assets/172641786/b8197f3b-9954-436d-a1b6-450fbae51b71)

**Reply by TactileDesign** [reply]
A note about the apparent 5V on the output of the drivers - when I grounded the scope to the power supply ground, the 5V went away - so it was just an artifact of the scope ground not being referenced properly...

**Comment by kriswilk** [comment]
I always recommend installing the [latest stable CircuitPython for QT Py RP2040](https://circuitpython.org/board/adafruit_qtpy_rp2040/), which is currently 9.0.5. The `lib` folder in this repo had files from the CircuitPython 8.x branch, which was the latest at the time I posted them. I've updated the repo just now to include the 9.x versions instead.

SO...first make sure you have CircuitPython 9.0.5 flashed onto the device. Then copy the new `lib` folder, overwriting all files. The `code.py` file hasn't changed. See if that makes any difference.

If you still have trouble, here's a long shot based on your photo: The 4-pin connectors on your boards don't all look like they are fully seated. If you look very closely at my photo of the assembled parts, you can see that the plugs seat deeply into the connectors.

![Screenshot 2024-06-18 211148](https://github.com/kriswilk/buzzah/assets/382436/c291049d-0725-4ce7-bbe5-e17bc6ba08ab)
![Screenshot 2024-06-18 211219](https://github.com/kriswilk/buzzah/assets/382436/de140d85-c68e-4f95-ace6-1cfdf227a5ae)

If a connector between the QT Py and the mux isn't fully seated, the mux may not be visible to the processor and you'd definitely get the kinds of errors you are seeing.

Like I said, a long shot, but worth a check.

**Reply by TactileDesign** [reply]
Kris,  Thanks again for all your suggestions.  To confirm, in order for the RP to complete its cycle, I need to have the MUX connected securely **and at least one drive board connected to the mux, but motors are optional?**  Is that correct?  I will give it another go tomorrow with all new downloads and tightened connections, and if still no luck, I will start from scratch with a new RP + new Mux + new driver board + alternate computer to load files.  Thanks!, Mark

**Reply by kriswilk** [reply]
The mux board must be connected, but the code should run without error/red light even if NO driver(s) at all are connected. Early in execution the code checks each port for a driver and will simply do nothing on any port that is unpopulated. But if the code cannot talk to the mux board, it will definitely fail.

As for motors, they shouldn't have to be connected to the drivers for the code to run either. The code merely sends instructions to the drivers, and the drivers apply output signals. If the drivers were configured to auto-tune the frequency there might be issues but in this case the drivers all run in open-loop mode with no feedback.

**Reply by TactileDesign** [reply, marked as answer]
Kris,  Everything is working perfectly now.  Thanks so much for your help yesterday.  For posterity and perhaps others who have my same problem in the future, here is what I did today:
1.  Removed all motors from the build I had yesterday that didnt work
2. Checked seating of cable between RP and MUX as Kris suggested.  Neither connection was solid.  Seated with substantial force and got extra travel - definitely a much better connection with the extra force
3. Tried system again (no driver boards) with same null result - red light flashing on RP
4. Downloaded the new LIB and CODE files Kris uploaded yesterday.  Downloaded 9.0.5 again.  Repeated Kris's set up procedure to load files on RP
5. RP stopped blinking red!
6. Attached a driver board with LRA - Buzzing commenced!  (Steps 4 through 6 took only about 5 minutes - very simple...) 
7. Adjusted amplitude settings in CODE using Mu.  Loaded CODE on to the RP while it was operating.  After a 1-2 second pause, started buzzing with more amplitude as per my new settings.  

The system now seems robust and I look forward to having some time tomorrow to look at the buzzer response with my scope...

My thanks again to Kris for all his help yesterday and for developing this system...

**Reply by kriswilk** [reply]
Glad to hear that you got it working and apologies for the out-of-date library files. I included copies of them in this repo for convenience (especially for less technically-minded builders). But as a consequence, I need to keep them synced with the CircuitPython library source.
